operators/pkg/controller/elasticsearch/certificates/transport/pod_secret.go

```diff
@@ -5,80 +5,216 @@
-
-			Labels: map[string]string{
-				// a label that allows us to list secrets of this type
-				LabelCertificateType: LabelCertificateTypeTransport,
-				// a label referencing the related pod so we can look up the pod from this secret
-				label.PodNameLabelName: pod.Name,
-				// a label showing which cluster this pod belongs to
-				label.ClusterNameLabelName: es.Name,
-			},
-		},
+	svcs []corev1.Service,
+	ca *certificates.CA,
+	rotationParams certificates.RotationParams,
+) error {
+	// verify that the secret contains a parsable private key, create if it does not exist
+	var privateKey *rsa.PrivateKey
+	needsNewPrivateKey := true
+	if privateKeyData, ok := secret.Data[PodKeyFileName(pod.Name)]; ok {
+		storedPrivateKey, err := certificates.ParsePEMPrivateKey(privateKeyData)
+		if err != nil {
+			log.Error(err, "Unable to parse stored private key", "pod", pod.Name)
+		} else {
+			needsNewPrivateKey = false
+			privateKey = storedPrivateKey
+		}
+	}
+
+	// if we need a new private key, generate it
+	if needsNewPrivateKey {
+		generatedPrivateKey, err := rsa.GenerateKey(cryptorand.Reader, 2048)
+		if err != nil {
+			return err
+		}
+
+		privateKey = generatedPrivateKey
+		secret.Data[PodKeyFileName(pod.Name)] = certificates.EncodePEMPrivateKey(*privateKey)
+	}
+
+	if shouldIssueNewCertificate(es, *secret, pod, privateKey, svcs, ca, rotationParams.RotateBefore) {
+		log.Info(
+			"Issuing new certificate",
+			"pod", pod.Name,
+		)
+
+		csr, err := x509.CreateCertificateRequest(cryptorand.Reader, &x509.CertificateRequest{}, privateKey)
+		if err != nil {
+			return err
+		}
+
+		// create a cert from the csr
+		parsedCSR, err := x509.ParseCertificateRequest(csr)
+		if err != nil {
+			return err
+		}
+
+		validatedCertificateTemplate, err := createValidatedCertificateTemplate(
+			pod, es, svcs, parsedCSR, rotationParams.Validity,
+		)
+		if err != nil {
+			return err
+		}
+		// sign the certificate
+		certData, err := ca.CreateCertificate(*validatedCertificateTemplate)
+		if err != nil {
+			return err
+		}
+
+		// store the issued certificate in a secret mounted into the pod
+		secret.Data[PodCertFileName(pod.Name)] = certificates.EncodePEMCert(certData, ca.Cert.Raw)
+	}
+
+	return nil
+}
+
+// shouldIssueNewCertificate returns true if we should issue a new certificate.
+//
+// Reasons for reissuing a certificate:
+// - no certificate yet
+// - certificate has the wrong format
+// - certificate is invalid or expired
+// - certificate SAN and IP does not match pod SAN and IP
+func shouldIssueNewCertificate(
+	es v1alpha1.Elasticsearch,
+	secret corev1.Secret,
+	pod corev1.Pod,
+	privateKey *rsa.PrivateKey,
+	svcs []corev1.Service,
+	ca *certificates.CA,
+	certReconcileBefore time.Duration,
+) bool {
+	certCommonName := buildCertificateCommonName(pod, es.Name, es.Namespace)
+
+	generalNames, err := buildGeneralNames(es, svcs, pod)
+	if err != nil {
+		log.Error(err, "Cannot create GeneralNames for the TLS certificate", "pod", pod.Name)
```

Comment:
Thinking about this #1168 (which standardizes on using namespace and name fields where appropriate for easy searching of logs), should we use `pod_name`? And add the namespace?